api/v1alpha1/clusterconfig_types.go

```diff
@@ -237,52 +241,78 @@ func (ExtraAPIServerCertSANs) VariableSchema() clusterv1.VariableSchema {
 	}
 }
 
-type ImageRegistries struct {
+type ImageCredentials struct {
+	// The Secret containing the registry credentials and CA certificate
+	// The Secret should have keys 'username', 'password' and 'ca.crt'
+	// This credentials Secret is not required for some registries, e.g. ECR.
 	// +optional
-	ImageRegistryCredentials ImageRegistryCredentials `json:"credentials,omitempty"`
+	SecretRef *corev1.ObjectReference `json:"secretRef,omitempty"`
 }
 
-func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
+func (ImageCredentials) VariableSchema() clusterv1.VariableSchema {
 	return clusterv1.VariableSchema{
 		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
-			Description: "Configuration for image registries.",
-			Type:        "object",
+			Type: "object",
 			Properties: map[string]clusterv1.JSONSchemaProps{
-				"credentials": ImageRegistryCredentials{}.VariableSchema().OpenAPIV3Schema,
+				"secretRef": {
+					Description: "The Secret containing the registry credentials. " +
+						"The Secret should have keys 'username', 'password' and 'ca.crt' " +
+						"This credentials Secret is not required for some registries, e.g. ECR.",
+					Type: "object",
+					Properties: map[string]clusterv1.JSONSchemaProps{
+						"name": {
+							Description: "The name of the Secret containing the registry credentials.",
+							Type:        "string",
+						},
+						"namespace": {
+							Description: "The namespace of the Secret containing the registry credentials. " +
+								"Defaults to the namespace of the KubeadmControlPlaneTemplate and KubeadmConfigTemplate" +
+								"that reference this variable.",
+							Type: "string",
+						},
+					},
+				},
 			},
 		},
 	}
 }
 
-type ImageRegistryCredentials []ImageRegistryCredentialsResource
+// GlobalImageRegistryMirror sets default mirror configuration for all the image registries.
+type GlobalImageRegistryMirror struct {
+	// Registry URL.
+	URL string `json:"url"`
 
-func (ImageRegistryCredentials) VariableSchema() clusterv1.VariableSchema {
-	resourceSchema := ImageRegistryCredentialsResource{}.VariableSchema().OpenAPIV3Schema
+	// Credentials and CA certificate for the image registry mirror
+	// +optional
+	Credentials *ImageCredentials `json:"credentials,omitempty"`
+}
 
+func (GlobalImageRegistryMirror) VariableSchema() clusterv1.VariableSchema {
 	return clusterv1.VariableSchema{
 		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
-			Description: "Image registry credentials to set up on all Nodes in the cluster. " +
-				"Enabling this will configure the Kubelets with " +
-				"https://kubernetes.io/docs/tasks/administer-cluster/kubelet-credential-provider/.",
-			Type:  "array",
-			Items: &resourceSchema,
+			Type: "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
+				"url": {
+					Description: "Registry mirror URL.",
+					Type:        "string",
```

Comment:
Let's add URI format to this so it will be validated via OpenAPI (you can see the valid formats at https://github.com/kubernetes/apiextensions-apiserver/blob/v0.28.5/pkg/apis/apiextensions/v1/types_jsonschema.go#L47-L72) although this would also allow absolute paths(see https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/url/url_test.go;l=672-729 for what would be considered valid in this case):

```suggestion
					Type:        "string",
					Format:      "uri",
```

Or a pattern:

```suggestion
					Type:        "string",
					Pattern:     "^https?://", // Consider something stricter here.
```

Or perhaps a combo is best which says make sure this is a URI but with http or https scheme?

```suggestion
					Type:        "string",
					Format:      "uri",
					Pattern:     "^https?://",
```

Reply:
fixed: https://github.com/d2iq-labs/capi-runtime-extensions/pull/292/commits/0bc7cad2fc346c19f60c672360d27dfa1ff6174c